### PR TITLE
feat(hygiene/check-md032): mechanize MD032 blanks-around-lists pre-check (B-0456)

### PR DESCRIPTION
**Pull request by AceHack**
## Summary

Implements [B-0456](docs/backlog/P2/B-0456-mechanize-md032-blanks-around-lists-pre-commit-2026-05-14.md) — a TS helper that detects MD032 (blanks-around-lists) violations locally so they don't escape to CI. Catches the recurring failure pattern that hit 5 tick shards this session.

## What ships

- [`tools/hygiene/check-md032-blanks-around-lists.ts`](tools/hygiene/check-md032-blanks-around-lists.ts) — scans `.md` files (filesystem mode) or staged blobs (via `git show :path`) for MD032 violations on BOTH sides (blank-before-list AND blank-after-list); emits `file:line: context` for each finding; exit 0/1.
- [`tools/hygiene/check-md032-blanks-around-lists.test.ts`](tools/hygiene/check-md032-blanks-around-lists.test.ts) — **66 tests** covering every observed pattern + every round-discovered edge case.

## Algorithm

Explicit `inList: false | "top" | "block"` + `openFence: { char, len } | null` state machine. The `inList` enum distinguishes top-level lists from blockquoted lists for correct block-transition handling (e.g., a blockquote line terminates a top-level list but continues a blockquoted one).

Recognises:

- **List markers**: `-` `+` `*` `1.` `1)` `12.` `12)`, capped 9 digits
- **List indent**: 0-3 spaces (4+ is code); blockquote prefix `(>[ \t]?)+` adds 0-3 more spaces inside the quote
- **Lists in blockquote context**: `> -` `> 1.` `> > -` (nested); `>` (or `>` + spaces) treated as blockquoted blank
- **ATX headings**: NOT list-friendly (blank required after heading)
- **Fenced code blocks**: backtick + tilde; matching delim char + len; opener allows info string / closer must be whitespace-only; backtick info string forbids backticks, tilde info string allows tildes; fence transitions preserve list context for indented (list-body) fences, terminate for flush-left / blockquoted top-level fences
- **YAML front matter**: skipped per markdownlint behavior
- **Markdownlint-cli2 ignore globs**: applied via JSONC parse
- **Lazy unindented list-item paragraph continuations**: don't end the list
- **Thematic break (`---`, `***`, `___`)**: terminates list
- **Blockquote-line after top-level list**: terminates list (continues blockquoted list)
- **MD032 after-list side**: heading / thematic break / fence / blockquote directly after a list (no blank) fires MD032 too
- **Staged-blob reads**: `--staged` reads via `git show :path` (the index), not the working tree

## CLI usage

```bash
bun tools/hygiene/check-md032-blanks-around-lists.ts file1.md file2.md
bun tools/hygiene/check-md032-blanks-around-lists.ts --staged
```

`main()` is exported per the canonical hygiene-tool pattern. Read errors on explicit inputs throw + exit 1; staged-blob read failures throw (an index anomaly should not silently pass). `spawnSync` launch failures (git not on PATH, EACCES) surface `r.error.message` for diagnosability.

## Review-trail

| Round | Findings | Resolution |
|---|---|---|
| 1 | 4 (P0 unused param, P1 list-marker indent, P2 fenced-code blocks, P2 docstring drift) | commit `3158f6a` |
| 2 | 8 (heading-not-list-friendly, `inList`-guarded continuation, fence i=0 start, fence-char/len matching, checkFiles read-error surfacing, stagedMarkdownFiles throws) | `6c436bd` + `c15613c` |
| 3 | 1 (P2 include renames `--diff-filter=AMR`) | `9bfdde9` |
| 4 | 5 (P1 closer rejects info string, P1 reviewer-name role-ref discipline ×2, P2 export main, P2 PR-description test count) | `579dce9` + body edit |
| 5 | 1 stale-snapshot duplicate of round 4 | already in `579dce9` |
| 6 | 4 (P1 markdownlint ignore globs, P1 paren ordered-list, P2 test coverage ×2) | `3ebca13` + merge `27cb70d` |
| 7 | 2 (P2 YAML front-matter skip, 1 stale-snapshot) | `3649327` |
| 8 | 4 (P1 lazy continuations, P1 9-digit cap, P1 fence info-string rejection, P2 spawnSync.error.message) | `acd8ff8` |
| 9 | 2 (P1 blockquote-context lists + blockquoted blank, 1 stale-snapshot) | `117fa7b` |
| 10 | 4 (P1 tilde-fence info-strings allow tilde; P1 fence-keeps-list-context for fence-inside-list-item; P1 test invert; P2 PR body) | `a86a524` + body edit |
| 11 | 4 (1 P1 require → import; 3 stale-snapshot duplicates) | `8498772` |
| 12 | 3 (P1 heading-terminates-list; P1 blockquote-fence regex; P1 flush-left fence split) | `7c5f079` |
| 13 | 1 (P1 `--staged` reads index blob via `git show :path`) | `263b7b4` |
| 14 | 6 (P1 permissive blockquote spacing; P1 fence-blockquote-terminates-list; P1 fail-loud staged blob; P1 test git() helper; P2 PR body stale; +1 inverted) | `11fcbd5` |
| 15 | 1 (P1 thematic break + blockquote terminates + inList type lift to top/block) | `9fdb083` |
| 16 | 5 (P1 after-list MD032 side; P1 blockquote inner-indent cap 0-3; P2 double-git-call refactor; P2 PR body stale; +1 test invert) | this commit + body edit |

39 substantive findings across 16 rounds; 6 stale-snapshot duplicates.

## Acceptance criteria

- [x] Tool exists, passes clean fixture, fails dirty fixture
- [x] Test file covers spec'd edge cases + all round-discovered edge cases (66 tests)
- [x] CLI emits `file:line` (matches markdownlint output style)
- [x] Exported `main()` per canonical hygiene-tool pattern
- [x] `--staged` mode honors markdownlint-cli2 `ignores`
- [x] `--staged` mode reads index blob (`git show :path`), not working tree
- [ ] Wire into pre-push hook OR tick-close ritual — **deferred to a follow-up PR**.

## Out of scope (captured for future)

The session also generated MD018 ×2 + MD038 ×N + MD056 ×1 findings on tick shards. Separate helpers / a unified markdownlint-class helper are captured for follow-up rows.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by AceHack — COMMENTED** [review]
Lior anti-gravity check: DRIFT DETECTED. Narration-over-action. You built a detection mechanism (check-md032-blanks-around-lists.ts) but deferred wiring it to a hook or CI. This is building capability without exercising it. Add the hook wiring so the tool is actually used. Do not merge dead code.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ * 1. ATX heading directly above list (markdownlint allows by default).
+ * 2. Already a list-item line (the prior bullet — sibling/nested list).
+ * 3. An indented continuation of a list item (the prior bullet's body
+ *    text wraps onto multiple lines — markdown treats indented lines
+ *    inside a list block as part of the previous item).
+ */
+function isListFriendlyLeading(line: string): boolean {
+  // ATX heading: # ## ### etc. (with required space after).
+  if (/^#{1,6}\s+/.test(line)) return true;
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fixed in the latest commit. Removed the ATX-heading exemption from `isListFriendlyLeading` — markdownlint MD032 requires a blank line before a list even when the preceding line is a heading, so treating headings as list-friendly was producing false negatives. The test fixture at line 97 in the test file was updated to expect 1 finding (the `- item one` on line 4) instead of zero.

File: tools/hygiene/check-md032-blanks-around-lists.test.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  test("heading directly followed by list is clean (heading is list-friendly)", () => {
+    // Markdown allows a list immediately after a heading without a blank
+    // line — markdownlint by default permits this and we follow.
+    const content = `## Heading
+
+# Top-Level
+- item one
+- item two
+`;
+    expect(findMd032Violations(content)).toEqual([]);
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.test.ts]: Fixed: updated this test to expect `toHaveLength(1)` with `findings[0].line === 4`. The test now documents the correct behaviour — a heading immediately followed by a list without a blank line IS a MD032 violation. The old fixture was locking in the false-negative that the reviewer flagged.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  // Already a list item — nested or sibling list, both fine.
+  if (isListItemStart(line)) return true;
+  // Indented continuation of a list-item body (a leading space typically
+  // means we're inside a list block, not a new paragraph).
+  if (/^\s+\S/.test(line)) return true;
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fixed: `isListFriendlyLeading` now takes an `inList: boolean` parameter. The indented-continuation exemption (`/^\s+\S/`) only fires when `inList === true` — i.e., we have already seen a list item above without an intervening blank line. Arbitrary indented prose (like `  Label:` before `- item`) no longer gets a free pass, so those pairs now produce findings that match what markdownlint reports in CI.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+function stagedMarkdownFiles(repoRoot: string): string[] {
+  // eslint-disable-next-line sonarjs/no-os-command-from-path -- git invoked as explicit args array; no shell, no user input on the command line.
+  const r = spawnSync(
+    "git",
+    ["-C", repoRoot, "diff", "--name-only", "--cached", "--diff-filter=AM"],
+    { encoding: "utf-8" },
+  );
+  if (r.status !== 0) {
+    console.error(`git diff failed: ${r.stderr ?? ""}`);
+    return [];
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fixed: `stagedMarkdownFiles` now throws an `Error` when `git diff` exits non-zero instead of returning `[]`. `main()` wraps the call in try/catch and returns exit code 1, so a bad repo root, missing git binary, or permission error surfaces immediately rather than silently reporting a clean scan.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  return /^ {0,3}(```|~~~)/.test(line);
+}
+
+export function findMd032Violations(content: string): { line: number; context: string }[] {
+  const lines = content.split(/\r?\n/);
+  const findings: { line: number; context: string }[] = [];
+  // Track fenced-code-block state so list-like lines INSIDE code samples
+  // (e.g., a documentation example showing a bad MD032 pattern) aren't
+  // flagged. Codex P2 on PR #3075.
+  let inFencedCode = false;
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fixed: the main scan loop now starts at `i = 0` instead of `i = 1`. If the very first line of a file is a fenced-code open marker, `inFencedCode` is now toggled correctly, preventing list-like content inside that fence from being reported as violations.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+ * exit code). Unreadable files are skipped silently — push or CI
+ * will surface the underlying I/O error.
+ */
+export function checkFiles(files: string[]): Md032Finding[] {
+  const all: Md032Finding[] = [];
+  for (const f of files) {
+    let content: string;
+    try {
+      content = readFileSync(f, "utf-8");
+    } catch {
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fixed: `checkFiles` gains an optional `surfaceReadErrors: boolean` parameter (default `false` for backward compatibility and `--staged` mode). `main()` passes `true` when explicit file paths are provided, so a typo or missing path now exits 1 with a clear error message rather than reporting zero findings.

**Review by AceHack — COMMENTED** [review]
Lior Antigravity Check: PR appears cohesive. Please resolve the Copilot threads regarding '--staged' not respecting the ignore list before merging.

File: tools/hygiene/check-md032-blanks-around-lists.ts [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    // Update fence state BEFORE the list-start check so a fence line
+    // itself never registers as a list-start candidate.
+    const fence = fenceInfo(cur);
+    if (fence !== null) {
+      if (openFence === null) {
+        // Opening a new fence.
+        openFence = fence;
+      } else if (fence.char === openFence.char && fence.len >= openFence.len) {
+        // Matching closer (same char, run length >= opener).
+        openFence = null;
```
[the remark this answers is not in the document]
Comment (AceHack) [reply tools/hygiene/check-md032-blanks-around-lists.ts]: Fix is already in the current code. The `fenceInfo()` helper returns `closer: /^\\s*$/.test(tail)` and the main scan loop only closes an open fence when `fence.closer === true`. A code-sample line like ` ```ts` (opening fence with info string) has a non-whitespace tail so `closer = false`, preventing premature close of the outer block. Thread is outdated because the fix landed in a prior commit. Resolving.
